Server/src/main/java/org/gluu/oxauth/service/CibaRequestsProcessorJob.java

```diff
@@ -0,0 +1,182 @@
+
+        try {
+            if (jobIsFree()) {
+                processImpl();
+                this.lastFinishedTime = System.currentTimeMillis();
+            } else {
+                log.trace("Starting conditions aren't reached for CIBA requestes processor");
+            }
+        } finally {
+            this.isActive.set(false);
+        }
+    }
+
+    /**
+     * Defines whether the job is still in process or it is free according to the time interval defined.
+     * @return True in case it is free to start a new process.
+     */
+    private boolean jobIsFree() {
+        int interval = appConfiguration.getBackchannelRequestsProcessorJobIntervalSec();
+        if (interval < 0) {
+            log.info("CIBA Requests processor timer is disabled.");
+            log.warn("CIBA Requests processor timer Interval (cleanServiceInterval in oxauth configuration) is negative which turns OFF internal clean up by the server. Please set it to positive value if you wish internal CIBA Requests processor up timer run.");
+            return false;
+        }
+
+        long timeDiffrence = System.currentTimeMillis() - this.lastFinishedTime;
+        return timeDiffrence >= interval * 1000;
+    }
+
+    /**
+     * Main process that process CIBA requests in cache.
+     */
+    public void processImpl() {
+        try {
+            List<CIBARequest> expiredRequests = cibaRequestService.loadExpiredByStatus(
+                    CIBAGrantUserAuthorization.AUTHORIZATION_PENDING);
+
+            for (CIBARequest expiredRequest : expiredRequests) {
+                CIBAGrant cibaGrant = authorizationGrantList.getCIBAGrant(expiredRequest.getAuthReqId());
+                if (cibaGrant != null) {
+                    executorService.execute(() ->
+                        processExpiredRequest(cibaGrant, expiredRequest.getAuthReqId())
+                    );
+                }
+                cibaRequestService.updateStatus(expiredRequest, CIBAGrantUserAuthorization.AUTHORIZATION_EXPIRED);
+            }
+        } catch (Exception e) {
+            log.error("Failed to process CIBA request from cache.", e);
+        }
+    }
+
+    /**
+     * Method responsible to process expired CIBA grants, set them as expired in cache
+     * and send callbacks to the client
+     * @param cibaGrant Object containing data related to the CIBA grant.
+     * @param authReqId Authentication request id.
+     */
+    private void processExpiredRequest(CIBAGrant cibaGrant, String authReqId) {
+        if (cibaGrant.getUserAuthorization() != CIBAGrantUserAuthorization.AUTHORIZATION_PENDING
+                && cibaGrant.getUserAuthorization() != CIBAGrantUserAuthorization.AUTHORIZATION_EXPIRED) {
+            return;
+        }
+        log.info("Authentication request id {} has expired", authReqId);
+
+        cibaGrant.setUserAuthorization(CIBAGrantUserAuthorization.AUTHORIZATION_EXPIRED);
+        cibaGrant.setTokensDelivered(false);
+        cibaGrant.save();
```

Comment:
Why do we save grant and still keep request in persistence? Shouldn't we remove both if it's expired?

Reply:
Ok, I think there are no specs about it, so I'll remove them then.